From 7c4d8c123ab17b9e325543344601aa136f995ec0 Mon Sep 17 00:00:00 2001
From: Nathan Ward <nward@searchlight.nz>
Date: Sun, 23 May 2021 21:33:39 +1200
Subject: [PATCH 1/3] Add parsing for octets coming in as integer or str values
 (i.e. from dictionaries with VALUE definitions for an attribute)

As part of this, don't overwrite "str"

[Canonical Notes: Cherry-picked to include changes to pyrad/tools.py only]
---
 pyrad/tools.py          | 17 +++++++++++------
 tests/testDictionary.py | 16 ++++++++++++++++
 2 files changed, 27 insertions(+), 6 deletions(-)

diff --git a/pyrad/tools.py b/pyrad/tools.py
index 3ddb33e..42019cb 100644
--- a/pyrad/tools.py
+++ b/pyrad/tools.py
@@ -17,15 +17,20 @@ def EncodeString(str):
         return str
 
 
-def EncodeOctets(str):
-    if len(str) > 253:
+def EncodeOctets(octetstring):
+    if len(octetstring) > 253:
         raise ValueError('Can only encode strings of <= 253 characters')
 
-    if str.startswith(b'0x'):
-        hexstring = str.split(b'0x')[1]
+    if type(octetstring) == bytes and octetstring.startswith(b'0x'):
+        hexstring = octetstring.split(b'0x')[1]
+        return binascii.unhexlify(hexstring)
+    elif type(octetstring) == str and octetstring.startswith('0x'):
+        hexstring = octetstring.split('0x')[1]
         return binascii.unhexlify(hexstring)
+    elif type(octetstring) == str and octetstring.isdecimal():
+        return struct.pack('>L',int(octetstring)).lstrip((b'\x00'))
     else:
-        return str
+        return octetstring
 
 
 def EncodeAddress(addr):
@@ -111,7 +116,7 @@ def EncodeAscendBinary(str):
 
     trailer = 8 * b'\x00'
 
-    result = b''.join((terms['family'], terms['action'], terms['direction'], b'\x00', 
+    result = b''.join((terms['family'], terms['action'], terms['direction'], b'\x00',
         terms['src'], terms['dst'], terms['srcl'], terms['dstl'], terms['proto'], b'\x00',
         terms['sport'], terms['dport'], terms['sportq'], terms['dportq'], b'\x00\x00', trailer))
     return result

From 8d4cf298a3a0e77cc149e054558ceffda69fad13 Mon Sep 17 00:00:00 2001
From: Nathan Ward <nward@searchlight.nz>
Date: Sun, 23 May 2021 22:16:48 +1200
Subject: [PATCH 3/3] Fix #152 and add tests for the same

---
 pyrad/tools.py     | 23 +++++++++++++++--------
 tests/testTools.py |  4 +++-
 2 files changed, 18 insertions(+), 9 deletions(-)

diff --git a/pyrad/tools.py b/pyrad/tools.py
index 42019cb..8425418 100644
--- a/pyrad/tools.py
+++ b/pyrad/tools.py
@@ -18,19 +18,26 @@ def EncodeString(str):
 
 
 def EncodeOctets(octetstring):
-    if len(octetstring) > 253:
+    # Check for max length of the hex encoded with 0x prefix, as a sanity check
+    if len(octetstring) > 508:
         raise ValueError('Can only encode strings of <= 253 characters')
 
-    if type(octetstring) == bytes and octetstring.startswith(b'0x'):
+    if isinstance(octetstring, bytes) and octetstring.startswith(b'0x'):
         hexstring = octetstring.split(b'0x')[1]
-        return binascii.unhexlify(hexstring)
-    elif type(octetstring) == str and octetstring.startswith('0x'):
+        encoded_octets = binascii.unhexlify(hexstring)
+    elif isinstance(octetstring, str) and octetstring.startswith('0x'):
         hexstring = octetstring.split('0x')[1]
-        return binascii.unhexlify(hexstring)
-    elif type(octetstring) == str and octetstring.isdecimal():
-        return struct.pack('>L',int(octetstring)).lstrip((b'\x00'))
+        encoded_octets = binascii.unhexlify(hexstring)
+    elif isinstance(octetstring, str) and octetstring.isdecimal():
+        encoded_octets = struct.pack('>L',int(octetstring)).lstrip((b'\x00'))
     else:
-        return octetstring
+        encoded_octets = octetstring
+
+    # Check for the encoded value being longer than 253 chars
+    if len(encoded_octets) > 253:
+        raise ValueError('Can only encode strings of <= 253 characters')
+
+    return encoded_octets
 
 
 def EncodeAddress(addr):
